File: tools/dropopts.py
#!/usr/bin/env python3
"""Delete command-line options, so that each reaches the error an unknown one does.

Usage:
    python3 tools/dropopts.py <file> -X -Y --name ...

An option the editor accepts and ignores is a lie, and an option whose whole
body is an error message is a branch that exists only to say no.  Both are
better expressed by the option not existing -- which this build already has a
path for, `mainerr(ME_UNKNOWN_OPTION)`, reached by anything the parser does not
recognise.  WHIM-GOAL.md's Phase 3 says which options, and why each.

A SHORT OPTION has up to two `case` labels, one in each of command_line_scan()'s
two switches: the one that reads the option, and the one that reads its argument
when it takes one.  A label that shares its body with other options (`case 'S':
case 'd': case 'T':`) goes alone and the body stays for the rest; a label that is
the last of its group goes with the body.  A whole group is refused if the case
before it falls through into it, because the body would vanish from under an
option that is staying.

A LONG OPTION is a link in an `if ... else if` chain, removed by brace matching.
THE CHAIN MUST STAY A CHAIN, and the first version of this tool did not keep it.
It asked whether the text before the removed link ended in `else` -- which it
never does, because the match had already consumed that `else` -- so removing
ANY link turned the next `else if` into a bare `if`.  `--clean`, `--noplugin` and
`--not-a-term` then each matched their own branch, failed every test of the
split chain after it, and reached `mainerr` anyway: three options broken for
thirty phases, because no harness passed them.  The removed link's own `else`
decides now, and tools/clicheck.py runs every option the parser has left.

Everything is bounded by command_line_scan()'s own text.  `case 'X':` occurs in
several switches in this file -- the normal-mode tables and get_c_indent() have
their own -- and a scan over the whole file finds the wrong one.  Everything
refuses a shape it does not recognise, because a partial removal here leaves an
option that parses and then runs something else's body.
"""

import re
import sys
from pathlib import Path

sys.path.insert(0, __file__.rsplit('/', 1)[0])
import cutil

LABEL = re.compile(r"^[ \t]*case '(.)':[ \t]*$")
DEFAULT = re.compile(r'^[ \t]*default:[ \t]*$')


def switches(fn):
    """The option switch and the argument switch, as (open, close) offsets in `fn`."""
    b = cutil.blank(fn)
    found = []
    for m in re.finditer(r'\bswitch \(c\)', b):
        o = b.index('{', m.end())
        found.append((o, cutil.match(fn, o, b)))
    if len(found) != 2:
        sys.exit('dropopts: expected the option switch and the argument switch, '
                 'found %d switches on c' % len(found))
    return found, b


def groups(lines, blines):
    """Each group of adjacent labels at the switch's own depth: (labels, body start, body end)."""
    depth, depths = 0, []
    for bl in blines:
        depths.append(depth)
        depth += bl.count('{') - bl.count('}')

    def is_label(k):
        return depths[k] == 0 and LABEL.match(lines[k])

    out, i = [], 0
    while i < len(lines):
        if not is_label(i):
            i += 1
            continue
        labels = []
        while i < len(lines) and is_label(i):
            labels.append(i)
            i += 1
        j = i
        while j < len(lines) and not (is_label(j) or
                                      (depths[j] == 0 and DEFAULT.match(lines[j]))):
            j += 1
        out.append((labels, i, j))
        i = j
    return out


def drop_short(fn, which, letters):
    """Remove `letters` from one switch; return (fn, the letters it held)."""
    sw, b = switches(fn)
    o, c = sw[which]
    lines = fn[o + 1:c].split('\n')
    gs = groups(lines, b[o + 1:c].split('\n'))
    remove, held = set(), set()
    for n, (labels, bs, be) in enumerate(gs):
        names = [LABEL.match(lines[k]).group(1) for k in labels]
        gone = [k for k, name in zip(labels, names) if name in letters]
        if not gone:
            continue
        held.update(LABEL.match(lines[k]).group(1) for k in gone)
        if len(gone) < len(labels):
            remove.update(gone)
            continue
        if n > 0:
            prev = [l.strip() for l in lines[gs[n - 1][1]:gs[n - 1][2]] if l.strip()]
            if prev and prev[-1] != 'break;':
                sys.exit("dropopts: -%s -- the case before it falls through into it, "
                         "so removing its body would change what that option runs"
                         % '/-'.join(names))
        tail = [l.strip() for l in lines[bs:be] if l.strip()]
        if not tail or tail[-1] != 'break;':
            sys.exit("dropopts: -%s -- its body does not end in break" % '/-'.join(names))
        remove.update(range(labels[0], be))
    kept = [l for k, l in enumerate(lines) if k not in remove]
    return fn[:o + 1] + '\n'.join(kept) + fn[c:], held


def drop_long(fn, name):
    """Remove the chain link comparing against "name", keeping the chain a chain."""
    sw, _ = switches(fn)
    o, c = sw[0]
    seg = fn[o + 1:c]
    pat = re.compile(r'^[ \t]*(else )?if \([^\n]*\("%s"\)[^\n]*\)\n[ \t]*\{'
                     % re.escape(name), re.M)
    ms = list(pat.finditer(seg))
    if len(ms) != 1:
        sys.exit('dropopts: --%s -- expected one branch in the option switch, found %d'
                 % (name, len(ms)))
    m = ms[0]
    b = cutil.blank(seg)
    ob = b.index('{', m.end() - 1)
    cb = cutil.match(seg, ob, b)
    end = cb + 1
    if seg[end:end + 1] == '\n':
        end += 1
    head, tail = seg[:m.start()], seg[end:]
    if not m.group(1):
        # The FIRST link: whatever followed it has to stop being an `else`.
        nxt = re.match(r'([ \t]*)else if\b', tail)
        if nxt:
            tail = nxt.group(1) + 'if' + tail[nxt.end():]
        elif re.match(r'[ \t]*else\b', tail):
            sys.exit('dropopts: --%s is the only test before a bare else' % name)
    return fn[:o + 1] + head + tail + fn[c:]


def main():
    if len(sys.argv) < 3:
        sys.exit(__doc__)
    path = Path(sys.argv[1])
    text = path.read_text(errors='surrogateescape')

    shorts, longs = [], []
    for a in sys.argv[2:]:
        if re.fullmatch(r'-[A-Za-z?]', a):
            shorts.append(a[1])
        elif re.fullmatch(r'--[a-z][a-z-]*', a):
            longs.append(a[2:])
        else:
            sys.exit('dropopts: not an option this can remove: %r' % a)

    span = cutil.find_definition(text, 'command_line_scan')
    if not span:
        sys.exit('dropopts: command_line_scan is not defined at file scope')
    a, z = span
    fn = text[a:z]

    fn, held = drop_short(fn, 0, set(shorts))
    missing = sorted(set(shorts) - held)
    if missing:
        sys.exit('dropopts: no `case` in the option switch for %s -- it has gone '
                 'already, or the switch has moved' % ' '.join('-' + c for c in missing))
    fn, in_args = drop_short(fn, 1, set(shorts))

    for n in longs:
        fn = drop_long(fn, n)

    # A long option that took an argument had a test in the argument switch,
    # and once its branch is gone that test compares against a name nothing
    # can produce.  An EMPTY one is dead whatever it names -- the one left
    # behind by --gui-dialog-file had been empty since the GUI went.
    sw, _ = switches(fn)
    o, c = sw[1]
    seg, empty = re.subn(r'^[ \t]*if \(argv\[-1\]\[2\] == \'.\'\)\n[ \t]*\{\n[ \t]*\}\n',
                         '', fn[o + 1:c], flags=re.M)
    fn = fn[:o + 1] + seg + fn[c:]

    path.write_text(text[:a] + fn + text[z:], errors='surrogateescape')
    print('  dropopts     %d short (%s), %d of them also in the argument switch; '
          '%d long (%s); %d empty argument test%s'
          % (len(shorts), ' '.join('-' + c for c in shorts), len(in_args),
             len(longs), ' '.join('--' + n for n in longs),
             empty, '' if empty == 1 else 's'))


if __name__ == '__main__':
    main()

File: tools/go/internal/cut/dropopts.go
package cut

import (
	"bytes"
	"fmt"
	"io"
	"regexp"
	"sort"
	"strings"

	"slimvim.local/tools/internal/cutil"
)

var (
	optLabel   = regexp.MustCompile(`^[ \t]*case '(.)':[ \t]*$`)
	optDefault = regexp.MustCompile(`^[ \t]*default:[ \t]*$`)
	switchOnC  = regexp.MustCompile(`\bswitch \(c\)`)
)

type optSwitch struct{ open, close int }

// Switches returns the option switch and the argument switch as offsets into
// fn, with the blanked copy the caller goes on using.
//
// EXACTLY TWO, asserted: command_line_scan has one switch that reads an
// option's letter and one that reads its argument, and a tool that took "the
// first" of some other number would edit whichever one happened to come first.
func Switches(fn []byte) ([]optSwitch, []byte, error) {
	b := cutil.Blank(fn)
	var found []optSwitch
	for _, m := range switchOnC.FindAllIndex(b, -1) {
		rel := bytes.IndexByte(b[m[1]:], '{')
		if rel < 0 {
			continue
		}
		o := m[1] + rel
		found = append(found, optSwitch{o, cutil.Match(b, o)})
	}
	if len(found) != 2 {
		return nil, nil, fmt.Errorf("dropopts: expected the option switch and the "+
			"argument switch, found %d switches on c", len(found))
	}
	return found, b, nil
}

type labelGroup struct {
	labels []int // line indices of the adjacent `case 'x':` labels
	bs, be int   // the body, as a half-open line range
}

// optGroups returns each run of adjacent labels at the switch's OWN depth,
// with the body that follows it.  The depth comes from the blanked lines, so a
// brace inside a string literal cannot move it.
func optGroups(lines, blines []string) []labelGroup {
	depth := 0
	depths := make([]int, len(blines))
	for i, bl := range blines {
		depths[i] = depth
		depth += strings.Count(bl, "{") - strings.Count(bl, "}")
	}
	isLabel := func(k int) bool {
		return k < len(lines) && depths[k] == 0 && optLabel.MatchString(lines[k])
	}
	var out []labelGroup
	for i := 0; i < len(lines); {
		if !isLabel(i) {
			i++
			continue
		}
		var labels []int
		for i < len(lines) && isLabel(i) {
			labels = append(labels, i)
			i++
		}
		j := i
		for j < len(lines) && !(isLabel(j) ||
			(depths[j] == 0 && optDefault.MatchString(lines[j]))) {
			j++
		}
		out = append(out, labelGroup{labels, i, j})
		i = j
	}
	return out
}

func stripped(lines []string) []string {
	var out []string
	for _, l := range lines {
		if s := strings.TrimSpace(l); s != "" {
			out = append(out, s)
		}
	}
	return out
}

// DropShort removes `letters` from one switch and reports which it held.
//
// The two refusals are the whole value of the tool.  A group whose PREVIOUS
// case falls through into it cannot lose its body, because that would change
// what the previous option runs; and a body that does not end in `break;` is
// one that falls through into the NEXT, which is the same hazard from the
// other side.
func DropShort(fn []byte, which int, letters map[string]bool) ([]byte, map[string]bool, error) {
	sw, b, err := Switches(fn)
	if err != nil {
		return nil, nil, err
	}
	o, c := sw[which].open, sw[which].close
	lines := strings.Split(string(fn[o+1:c]), "\n")
	blines := strings.Split(string(b[o+1:c]), "\n")
	gs := optGroups(lines, blines)

	remove := map[int]bool{}
	held := map[string]bool{}
	for n, g := range gs {
		var names []string
		for _, k := range g.labels {
			names = append(names, optLabel.FindStringSubmatch(lines[k])[1])
		}
		var gone []int
		for i, k := range g.labels {
			if letters[names[i]] {
				gone = append(gone, k)
			}
		}
		if len(gone) == 0 {
			continue
		}
		for _, k := range gone {
			held[optLabel.FindStringSubmatch(lines[k])[1]] = true
		}
		if len(gone) < len(g.labels) {
			for _, k := range gone {
				remove[k] = true
			}
			continue
		}
		if n > 0 {
			prev := stripped(lines[gs[n-1].bs:gs[n-1].be])
			if len(prev) > 0 && prev[len(prev)-1] != "break;" {
				return nil, nil, fmt.Errorf("dropopts: -%s -- the case before it falls "+
					"through into it, so removing its body would change what that "+
					"option runs", strings.Join(names, "/-"))
			}
		}
		tail := stripped(lines[g.bs:g.be])
		if len(tail) == 0 || tail[len(tail)-1] != "break;" {
			return nil, nil, fmt.Errorf("dropopts: -%s -- its body does not end in break",
				strings.Join(names, "/-"))
		}
		for k := g.labels[0]; k < g.be; k++ {
			remove[k] = true
		}
	}

	var kept []string
	for k, l := range lines {
		if !remove[k] {
			kept = append(kept, l)
		}
	}
	out := make([]byte, 0, len(fn))
	out = append(out, fn[:o+1]...)
	out = append(out, strings.Join(kept, "\n")...)
	return append(out, fn[c:]...), held, nil
}

// DropLong removes the chain link comparing against "name", KEEPING THE CHAIN
// A CHAIN: if the link removed was the first, whatever followed it has to stop
// being an `else`.
func DropLong(fn []byte, name string) ([]byte, error) {
	sw, _, err := Switches(fn)
	if err != nil {
		return nil, err
	}
	o, c := sw[0].open, sw[0].close
	seg := fn[o+1 : c]
	pat := regexp.MustCompile(fmt.Sprintf(
		`(?m)^[ \t]*(else )?if \([^\n]*\("%s"\)[^\n]*\)\n[ \t]*\{`, regexp.QuoteMeta(name)))
	ms := pat.FindAllSubmatchIndex(seg, -1)
	if len(ms) != 1 {
		return nil, fmt.Errorf("dropopts: --%s -- expected one branch in the option "+
			"switch, found %d", name, len(ms))
	}
	m := ms[0]
	b := cutil.Blank(seg)
	ob := m[1] - 1 + bytes.IndexByte(b[m[1]-1:], '{')
	cb := cutil.Match(b, ob)
	if cb < 0 {
		return nil, fmt.Errorf("dropopts: --%s -- its branch is unbalanced", name)
	}
	end := cb + 1
	if end < len(seg) && seg[end] == '\n' {
		end++
	}
	head, tail := seg[:m[0]], seg[end:]
	if m[2] < 0 { // no `else ` -- this is the FIRST link
		if nxt := regexp.MustCompile(`^([ \t]*)else if\b`).FindSubmatchIndex(tail); nxt != nil {
			var t []byte
			t = append(t, tail[nxt[2]:nxt[3]]...)
			t = append(t, "if"...)
			tail = append(t, tail[nxt[1]:]...)
		} else if regexp.MustCompile(`^[ \t]*else\b`).Match(tail) {
			return nil, fmt.Errorf("dropopts: --%s is the only test before a bare else", name)
		}
	}
	out := make([]byte, 0, len(fn))
	out = append(out, fn[:o+1]...)
	out = append(out, head...)
	out = append(out, tail...)
	return append(out, fn[c:]...), nil
}

var (
	shortOpt  = regexp.MustCompile(`^-[A-Za-z?]$`)
	longOpt   = regexp.MustCompile(`^--[a-z][a-z-]*$`)
	emptyTest = regexp.MustCompile(`(?m)^[ \t]*if \(argv\[-1\]\[2\] == '.'\)\n[ \t]*\{\n[ \t]*\}\n`)
)

// DropOpts is the command line: remove these options from command_line_scan.
func DropOpts(text []byte, args []string, w io.Writer) ([]byte, error) {
	var shorts, longs []string
	for _, a := range args {
		switch {
		case shortOpt.MatchString(a):
			shorts = append(shorts, a[1:])
		case longOpt.MatchString(a):
			longs = append(longs, a[2:])
		default:
			return nil, fmt.Errorf("dropopts: not an option this can remove: %s", cutil.PyRepr(a))
		}
	}

	blanked := cutil.Blank(text)
	a, z, ok := cutil.FindDefinition(text, blanked, "command_line_scan")
	if !ok {
		return nil, fmt.Errorf("dropopts: command_line_scan is not defined at file scope")
	}
	fn := text[a:z]

	want := map[string]bool{}
	for _, s := range shorts {
		want[s] = true
	}
	fn, held, err := DropShort(fn, 0, want)
	if err != nil {
		return nil, err
	}
	var missing []string
	for _, s := range shorts {
		if !held[s] {
			missing = append(missing, "-"+s)
		}
	}
	if len(missing) > 0 {
		sort.Strings(missing)
		return nil, fmt.Errorf("dropopts: no `case` in the option switch for %s -- it has "+
			"gone already, or the switch has moved", strings.Join(missing, " "))
	}
	fn, inArgs, err := DropShort(fn, 1, want)
	if err != nil {
		return nil, err
	}
	for _, n := range longs {
		if fn, err = DropLong(fn, n); err != nil {
			return nil, err
		}
	}

	// A long option that took an argument had a test in the argument switch,
	// and once its branch is gone that test compares against a name nothing
	// can produce.  An EMPTY one is dead whatever it names -- the one left
	// behind by --gui-dialog-file had been empty since the GUI went.
	sw, _, err := Switches(fn)
	if err != nil {
		return nil, err
	}
	o, c := sw[1].open, sw[1].close
	seg := fn[o+1 : c]
	empty := len(emptyTest.FindAll(seg, -1))
	seg = emptyTest.ReplaceAll(seg, nil)
	var rebuilt []byte
	rebuilt = append(rebuilt, fn[:o+1]...)
	rebuilt = append(rebuilt, seg...)
	fn = append(rebuilt, fn[c:]...)

	var dashed, ddashed []string
	for _, s := range shorts {
		dashed = append(dashed, "-"+s)
	}
	for _, n := range longs {
		ddashed = append(ddashed, "--"+n)
	}
	plural := "s"
	if empty == 1 {
		plural = ""
	}
	fmt.Fprintf(w, "  dropopts     %d short (%s), %d of them also in the argument switch; "+
		"%d long (%s); %d empty argument test%s\n",
		len(shorts), strings.Join(dashed, " "), len(inArgs),
		len(longs), strings.Join(ddashed, " "), empty, plural)

	out := make([]byte, 0, len(text))
	out = append(out, text[:a]...)
	out = append(out, fn...)
	return append(out, text[z:]...), nil
}
